File: apps/nest/storage/factory.go
package storage

import (
	"fmt"

	"github.com/koyeo/nest/config"
)

// NewFromCredential creates an ObjectStorage from a decrypted StorageCredential.
func NewFromCredential(cred *config.StorageCredential) (ObjectStorage, error) {
	switch cred.Provider {
	case "oss":
		if cred.Endpoint == "" {
			return nil, fmt.Errorf("OSS requires 'endpoint'")
		}
		return NewOSSStorage(cred.Endpoint, cred.AccessKeyID, cred.AccessKeySecret, cred.BucketName)
	case "s3":
		if cred.Region == "" {
			return nil, fmt.Errorf("S3 requires 'region'")
		}
		return NewS3Storage(cred.Region, cred.AccessKeyID, cred.AccessKeySecret, cred.BucketName, cred.Endpoint)
	default:
		return nil, fmt.Errorf("unsupported provider: '%s' (use 'oss' or 's3')", cred.Provider)
	}
}

File: apps/ship/src/storage/factory.ts
import type { StorageCredential } from "./provider.js";
import { OssStorage } from "./oss.js";
import { S3Storage } from "./s3.js";
import type { ObjectStorage } from "./storage.js";

/** Build a client from a decrypted credential. */
export function newStorage(cred: StorageCredential): ObjectStorage {
  switch (cred.provider) {
    case "oss":
      if (cred.endpoint === "") {
        throw new Error("OSS requires 'endpoint'");
      }
      return new OssStorage(cred.endpoint, cred.access_key_id, cred.access_key_secret, cred.bucket);
    case "s3":
      if (cred.region === "") {
        throw new Error("S3 requires 'region'");
      }
      return new S3Storage(cred.region, cred.access_key_id, cred.access_key_secret, cred.bucket, cred.endpoint);
  }
}
